File: src/player/PlayerPhysics.ts
import * as THREE from 'three';
import { PlayerConfig } from './PlayerConfig';
import { getBlockMovementProperties } from './BlockMovementProperties';
import type { PlayerCollision } from './PlayerCollision';
import type { PlayerInputSnapshot } from './PlayerInput';
import type { PlayerDamage } from './PlayerDamage';
import type { PlayerHunger } from './PlayerHunger';
import type { MovementState, PlayerLandedEvent } from './PlayerState';
import type { ChunkManager, CollisionAvailability } from '../world/ChunkManager';

export class PlayerPhysics {
  position = new THREE.Vector3(0, 80, 0);
  previousPosition = new THREE.Vector3(0, 80, 0);
  velocity = new THREE.Vector3();

  grounded = false;
  movementState: MovementState = 'idle';
  sprinting = false;
  sneaking = false;
  crawling = false;
  sitting = false;
  inWater = false;
  deepWater = false;
  inLava = false;
  lavaSubmersion = 0;
  blockedByStreaming = false;
  collisionAvailability: CollisionAvailability = 'READY';

  fallDistance = 0;
  private fallStartY: number | null = null;
  private justJumped = false;
  private lastJumpSource: 'input' | 'auto' | null = null;

  private onLandedCallbacks: Array<(evt: PlayerLandedEvent) => void> = [];

  constructor(
    private collision: PlayerCollision,
    private chunks: ChunkManager,
  ) {}

  onLanded(cb: (evt: PlayerLandedEvent) => void): () => void {
    this.onLandedCallbacks.push(cb);
    return () => {
      this.onLandedCallbacks = this.onLandedCallbacks.filter((c) => c !== cb);
    };
  }

  get currentHeight(): number {
    if (this.sitting) return PlayerConfig.dimensions.sittingHeight;
    if (this.crawling) return PlayerConfig.dimensions.crawlingHeight;
    if (this.sneaking) return PlayerConfig.dimensions.sneakingHeight;
    return PlayerConfig.dimensions.standingHeight;
  }

  get currentEyeHeight(): number {
    if (this.sitting) return PlayerConfig.dimensions.sittingEye;
    if (this.crawling) return PlayerConfig.dimensions.crawlingEye;
    if (this.sneaking) return PlayerConfig.dimensions.sneakingEye;
    return PlayerConfig.dimensions.standingEye;
  }

  get wasJustJumped(): boolean {
    return this.justJumped;
  }

  get jumpSource(): 'input' | 'auto' | null {
    return this.lastJumpSource;
  }

  /**
   * Executes a jump matching Minecraft Java Edition impulse physics.
   */
  private executeJump(
    hungerSystem: PlayerHunger,
    source: 'input' | 'auto',
    lookYaw: number,
  ): void {
    this.velocity.y = PlayerConfig.movement.jumpVelocity; // 0.42 blocks/tick
    if (this.sprinting) {
      const sinYaw = Math.sin(lookYaw);
      const cosYaw = Math.cos(lookYaw);
      this.velocity.x += -sinYaw * PlayerConfig.movement.sprintJumpForwardBoost;
      this.velocity.z += -cosYaw * PlayerConfig.movement.sprintJumpForwardBoost;
    }
    this.grounded = false;
    this.justJumped = true;
    this.lastJumpSource = source;
    hungerSystem.recordJump(this.sprinting);
  }

  /**
   * Deterministic 20 Hz simulation tick matching Minecraft Java Edition.
   */
  simulateTick(
    input: PlayerInputSnapshot,
    lookYaw: number,
    damageSystem: PlayerDamage,
    hungerSystem: PlayerHunger,
  ): void {
    this.justJumped = false;
    this.previousPosition.copy(this.position);

    if (damageSystem.isDead) {
      this.movementState = 'dead';
      this.velocity.set(0, 0, 0);
      return;
    }

    this.blockedByStreaming = false;
    this.collisionAvailability = 'READY';

    // A restored/player position must itself be backed by authoritative voxel
    // data before pose or physics logic interprets missing space as Air.
    const currentBox = this.collision.getPlayerAABB(
      this.position,
      PlayerConfig.dimensions.width,
      PlayerConfig.dimensions.standingHeight,
    );
    const currentAvailability = this.collision.getCollisionAvailabilityForRegion(
      currentBox.minX,
      currentBox.minY,
      currentBox.minZ,
      currentBox.maxX,
      currentBox.maxY,
      currentBox.maxZ,
    );
    if (currentAvailability !== 'READY') {
      this.blockedByStreaming = true;
      this.collisionAvailability = currentAvailability;
      this.velocity.set(0, 0, 0);
      this.sprinting = false;
      this.grounded = false;
      this.movementState = 'idle';
      return;
    }

    // 1. Stand up from sitting on movement key
    if (this.sitting && (input.forward || input.backward || input.left || input.right || input.jumpHeld)) {
      this.sitting = false;
    }

    // 2. Pose & Ceiling Clearance Verification (Crawl / Sneak / Stand)
    const canStand = this.collision.canStandUp(this.position);
    const canSneak = this.collision.canSneakUp(this.position);

    if (!canStand) {
      // Under low ceiling: force appropriate low space pose
      if (!canSneak) {
        this.crawling = true;
        this.sneaking = false;
      } else {
        this.sneaking = true;
        this.crawling = false;
      }
    } else {
      // Sufficient standing headroom available
      this.crawling = false;
      if (input.sneakHeld && !this.sitting) {
        this.sneaking = true;
      } else {
        this.sneaking = false;
      }
    }

    // 3. Sprint state verification with hunger requirement
    const wantsSprint = input.sprintHeld || input.sprintPressed;
    if (
      wantsSprint &&
      input.forward &&
      !this.sneaking &&
      !this.crawling &&
      !this.sitting &&
      hungerSystem.canSprint
    ) {
      this.sprinting = true;
    } else if (!input.forward || this.sneaking || this.crawling || this.sitting || !hungerSystem.canSprint) {
      this.sprinting = false;
    }

    // 4. Fluid state determination
    const height = this.currentHeight;
    this.inLava = this.chunks.isBodyInLava(this.position.x, this.position.y, this.position.z, height);
    this.inWater = !this.inLava && this.chunks.isBodyInWater(this.position.x, this.position.y, this.position.z, height);

    const waistY = this.position.y + height * 0.42;
    this.deepWater =
      this.inWater &&
      this.chunks.isWaterAt(Math.floor(this.position.x), Math.floor(waistY), Math.floor(this.position.z));

    if (this.inLava) {
      this.lavaSubmersion = Math.min(1, this.lavaSubmersion + 0.05 * 2.2);
    } else {
      this.lavaSubmersion = Math.max(0, this.lavaSubmersion - 0.05 * 3.5);
    }

    // 5. Horizontal wish direction
    const sinYaw = Math.sin(lookYaw);
    const cosYaw = Math.cos(lookYaw);
    const forward = new THREE.Vector3(-sinYaw, 0, -cosYaw);
    const right = new THREE.Vector3(cosYaw, 0, -sinYaw);

    const wishDir = new THREE.Vector3();
    if (!this.sitting) {
      if (input.forward) wishDir.add(forward);
      if (input.backward) wishDir.sub(forward);
      if (input.right) wishDir.add(right);
      if (input.left) wishDir.sub(right);

      if (input.analogX !== 0 || input.analogZ !== 0) {
        wishDir.addScaledVector(right, input.analogX);
        wishDir.addScaledVector(forward, -input.analogZ);
      }

      if (wishDir.lengthSq() > 1.0) {
        wishDir.normalize();
      }
    }

    // 6. Minecraft Java Acceleration & Friction parameters
    const groundBlock = this.collision.getGroundBlock(this.position);
    const blockProps = getBlockMovementProperties(groundBlock);
    const groundFriction = blockProps.friction; // 0.546 for standard block (S=0.6)
    const airFriction = PlayerConfig.movement.airFriction; // 0.91

    let accel = 0;
    if (this.grounded) {
      // Exact Java formula: f1 = 0.16277136 / (groundFriction^3)
      const q = 0.16277136 / Math.pow(groundFriction, 3);
      let factor: number = PlayerConfig.movement.walkAccelerationFactor; // 0.10
      if (this.sitting) {
        factor = 0;
      } else if (this.crawling) {
        factor *= PlayerConfig.movement.crawlMultiplier; // 0.3
      } else if (this.sneaking) {
        factor *= PlayerConfig.movement.sneakMultiplier; // 0.3
      } else if (this.sprinting) {
        factor *= PlayerConfig.movement.sprintMultiplier; // 1.3
      }
      accel = factor * q * blockProps.accelerationMultiplier;
    } else {
      if (this.sprinting) {
        accel = PlayerConfig.movement.airAccelerationSprint; // 0.026
      } else if (this.sneaking || this.crawling) {
        accel = PlayerConfig.movement.airAccelerationSneak; // 0.006
      } else {
        accel = PlayerConfig.movement.airAccelerationWalk; // 0.020
      }
    }

    if (this.deepWater) {
      accel *= PlayerConfig.movement.waterSpeedMultiplier;
    } else if (this.inLava) {
      accel *= PlayerConfig.movement.lavaSpeedMultiplier;
    }

    // Add acceleration to horizontal velocity
    if (wishDir.lengthSq() > 0) {
      this.velocity.x += wishDir.x * accel;
      this.velocity.z += wishDir.z * accel;
    }

    // 7. Jump Logic (Manual Jump + Holding Space + Auto-Jump)
    const canJump = this.grounded && !this.sitting && !this.deepWater && !this.inLava;
    if (canJump && (input.jumpPressed || input.jumpHeld)) {
      this.executeJump(hungerSystem, 'input', lookYaw);
    } else if (
      canJump &&
      !this.justJumped &&
      PlayerConfig.movement.autoJumpEnabled &&
      !this.sneaking &&
      !this.crawling &&
      wishDir.lengthSq() > 0.01
    ) {
      // Auto-jump candidate detection
      const autoJumpCheck = this.collision.checkAutoJumpCandidate(
        this.position,
        wishDir,
        PlayerConfig.dimensions.width,
        PlayerConfig.dimensions.standingHeight,
      );
      if (autoJumpCheck.canAutoJump) {
        this.executeJump(hungerSystem, 'auto', lookYaw);
      }
    }

    // 8. Vertical Motion & Fluid Mechanics
    if (this.inLava) {
      if (input.jumpHeld && !this.sitting) this.velocity.y += 0.08;
      if (this.sneaking || this.crawling) this.velocity.y -= 0.08;
      this.grounded = false;
    } else if (this.deepWater) {
      if (input.jumpHeld && !this.sitting) this.velocity.y += 0.09;
      if (this.sneaking || this.crawling) this.velocity.y -= 0.08;
      this.grounded = false;
    }

    // 9. Fall Tracking
    if (!this.grounded) {
      if (this.fallStartY === null) {
        this.fallStartY = this.position.y;
      }
    }

    // 10. Collision Resolution with swept AABB + step-up + sneak edge protection
    const collisionResult = this.collision.resolveMovement(
      this.position,
      this.velocity,
      height,
      PlayerConfig.dimensions.width,
      (this.sneaking || this.crawling) && this.grounded,
    );

    const prevGrounded = this.grounded;
    if (collisionResult.blockedByStreaming) {
      this.blockedByStreaming = true;
      this.collisionAvailability = collisionResult.collisionAvailability;
      this.velocity.set(0, 0, 0);
      // Preserve a previously verified grounded state because the position did
      // not move. Never manufacture grounded=true from an unresolved region.
      this.grounded = prevGrounded;
      this.sprinting = false;
      this.movementState = 'idle';
      return;
    }
    this.collisionAvailability = collisionResult.collisionAvailability;
    this.grounded = collisionResult.onGround;

    // 11. Post-Movement Vertical Gravity & Horizontal Friction Damping
    if (this.inLava) {
      this.velocity.y -= 0.02;
      this.velocity.y *= 0.85;
    } else if (this.deepWater) {
      this.velocity.y -= 0.02;
      this.velocity.y *= 0.88;
    } else if (!this.grounded) {
      this.velocity.y -= PlayerConfig.movement.gravity;
      this.velocity.y *= PlayerConfig.movement.verticalDrag;
    } else {
      this.velocity.y = 0;
    }

    const friction = this.grounded ? groundFriction : airFriction;
    this.velocity.x *= friction;
    this.velocity.z *= friction;

    // 12. Landing Logic & Fall Damage
    if (!prevGrounded && this.grounded) {
      let fallDist = 0;
      if (this.fallStartY !== null) {
        fallDist = Math.max(0, this.fallStartY - this.position.y);
      }
      this.fallStartY = null;
      this.fallDistance = fallDist;

      let damageTaken = 0;
      if (fallDist > PlayerConfig.damage.safeFallDistance && !this.inWater && !this.inLava) {
        damageTaken = Math.floor(fallDist - PlayerConfig.damage.safeFallDistance);
        if (damageTaken > 0) {
          damageSystem.processDamage(
            {
              amount: damageTaken,
              source: 'fall',
              bypassCooldown: true,
            },
            this.position,
            this.velocity,
          );
        }
      }

      if (fallDist > 0.5) {
        const landEvent: PlayerLandedEvent = {
          fallDistance: fallDist,
          landingVelocityY: this.velocity.y,
          surfaceBlock: collisionResult.groundBlock,
          wasSprinting: this.sprinting,
          damageTaken,
        };

        for (const cb of this.onLandedCallbacks) {
          cb(landEvent);
        }
      }
    }

    // 13. Distance-based Hunger Exhaustion
    const movedHoriz = Math.hypot(
      this.position.x - this.previousPosition.x,
      this.position.z - this.previousPosition.z,
    );

    if (this.sprinting && movedHoriz > 0) {
      hungerSystem.recordSprintDistance(movedHoriz);
    } else if (this.deepWater && movedHoriz > 0) {
      hungerSystem.recordSwimDistance(movedHoriz);
    }

    // 14. Determine Movement State
    const isMovingHoriz = movedHoriz > 0.001;
    if (this.inWater && this.deepWater) {
      this.movementState = 'swimming';
    } else if (this.crawling) {
      this.movementState = 'crawling';
    } else if (this.sneaking) {
      this.movementState = 'sneaking';
    } else if (this.justJumped) {
      this.movementState = 'jumping';
    } else if (!this.grounded && this.velocity.y < -0.1) {
      this.movementState = 'falling';
    } else if (this.sprinting && isMovingHoriz) {
      this.movementState = 'sprinting';
    } else if (isMovingHoriz) {
      this.movementState = 'walking';
    } else {
      this.movementState = 'idle';
    }
  }

  teleport(x: number, y: number, z: number): void {
    this.position.set(x, y, z);
    this.previousPosition.set(x, y, z);
    this.velocity.set(0, 0, 0);
    this.fallStartY = null;
    this.fallDistance = 0;
  }
}

File: Assets/VYTHERA/Player/Physics/PlayerPhysics.cs
using System;
using UnityEngine;
using VYTHERA.Player.Collision;
using VYTHERA.Voxel.Data;
using VYTHERA.World.Streaming;

namespace VYTHERA.Player.Physics
{
    public struct PlayerInputSnapshot
    {
        public bool Forward;
        public bool Backward;
        public bool Left;
        public bool Right;
        public bool JumpPressed;
        public bool JumpHeld;
        public bool SprintHeld;
        public bool SneakHeld;
        public float AnalogX;
        public float AnalogZ;
    }

    public sealed class PlayerPhysics : MonoBehaviour
    {
        [Header("State")]
        public Vector3 PreviousPosition = new Vector3(0, 80, 0);
        public Vector3 Position = new Vector3(0, 80, 0);
        public Vector3 Velocity;
        public bool Grounded;
        public bool Sprinting;
        public bool Sneaking;
        public bool Crawling;
        public bool InWater;
        public bool InLava;

        public float FallDistance;
        private float? _fallStartY;

        private PlayerCollision _collision;
        private ChunkManager _chunks;
        private bool _jumpArmed = true;

        public float CurrentHeight => Sneaking ? PlayerConfig.Dimensions.SneakingHeight : (Crawling ? PlayerConfig.Dimensions.CrawlingHeight : PlayerConfig.Dimensions.StandingHeight);
        public float CurrentEyeHeight => Sneaking ? PlayerConfig.Dimensions.SneakingEye : (Crawling ? PlayerConfig.Dimensions.CrawlingEye : PlayerConfig.Dimensions.StandingEye);

        public Vector3 InterpolatedPosition =>
            Vector3.Lerp(PreviousPosition, Position, Timing.FixedTickManager.Alpha);

        public event Action<float, byte> OnLanded;

        private void Update()
        {
            transform.position = InterpolatedPosition;
        }

        public void Initialize(ChunkManager chunks)
        {
            _chunks = chunks;
            _collision = new PlayerCollision(chunks);
            Position = transform.position;
            PreviousPosition = transform.position;
        }

        public void SimulateTick(PlayerInputSnapshot input, float lookYaw)
        {
            PreviousPosition = Position;
            // 1. Headroom verification
            bool canStand = !_collision.IsBoxBlocked(Position, PlayerConfig.Dimensions.Width, PlayerConfig.Dimensions.StandingHeight);
            bool canSneak = !_collision.IsBoxBlocked(Position, PlayerConfig.Dimensions.Width, PlayerConfig.Dimensions.SneakingHeight);

            if (!canStand)
            {
                if (!canSneak) { Crawling = true; Sneaking = false; }
                else { Sneaking = true; Crawling = false; }
            }
            else
            {
                Crawling = false;
                Sneaking = input.SneakHeld;
            }

            // 2. Sprint check
            if (input.SprintHeld && input.Forward && !Sneaking && !Crawling)
            {
                Sprinting = true;
            }
            else if (!input.Forward || Sneaking || Crawling)
            {
                Sprinting = false;
            }

            // 3. Fluid check
            byte bodyBlock = _chunks != null ? _chunks.GetBlock((int)MathF.Floor(Position.x), (int)MathF.Floor(Position.y + 0.5f), (int)MathF.Floor(Position.z)) : (byte)BlockType.Air;
            InWater = bodyBlock == (byte)BlockType.Water;
            InLava = bodyBlock == (byte)BlockType.Lava;

            // 4. Wish direction
            float sinYaw = MathF.Sin(lookYaw * Mathf.Deg2Rad);
            float cosYaw = MathF.Cos(lookYaw * Mathf.Deg2Rad);
            Vector3 forward = new Vector3(sinYaw, 0, cosYaw);
            Vector3 right = new Vector3(cosYaw, 0, -sinYaw);

            Vector3 wishDir = Vector3.zero;
            if (input.Forward) wishDir += forward;
            if (input.Backward) wishDir -= forward;
            if (input.Right) wishDir += right;
            if (input.Left) wishDir -= right;

            if (input.AnalogX != 0 || input.AnalogZ != 0)
            {
                wishDir += right * input.AnalogX + forward * input.AnalogZ;
            }

            if (wishDir.sqrMagnitude > 1f) wishDir.Normalize();

            // 5. Acceleration & friction
            float groundFriction = PlayerConfig.Movement.GroundFriction;
            float airFriction = PlayerConfig.Movement.AirFriction;

            float accel;
            if (Grounded)
            {
                float factor = PlayerConfig.Movement.WalkAccelerationFactor;
                if (Sprinting) factor *= PlayerConfig.Movement.SprintMultiplier;
                else if (Sneaking) factor *= PlayerConfig.Movement.SneakMultiplier;
                else if (Crawling) factor *= PlayerConfig.Movement.CrawlMultiplier;

                float q = 0.16277136f / MathF.Pow(groundFriction, 3);
                accel = factor * q;
            }
            else
            {
                accel = Sprinting ? PlayerConfig.Movement.AirAccelerationSprint : (Sneaking ? PlayerConfig.Movement.AirAccelerationSneak : PlayerConfig.Movement.AirAccelerationWalk);
            }

            if (InWater) accel *= PlayerConfig.Movement.WaterSpeedMultiplier;
            else if (InLava) accel *= PlayerConfig.Movement.LavaSpeedMultiplier;

            if (wishDir.sqrMagnitude > 0)
            {
                Velocity.x += wishDir.x * accel;
                Velocity.z += wishDir.z * accel;
            }

            // 6. Jump
            // Re-arm jump only when grounded and jump input is released
            if (Grounded && !input.JumpHeld && !input.JumpPressed)
            {
                _jumpArmed = true;
            }

            // Jump triggers ONLY on a fresh press when grounded and armed
            if (Grounded && _jumpArmed && input.JumpPressed && !InWater && !InLava)
            {
                Velocity.y = PlayerConfig.Movement.JumpVelocity;
                if (Sprinting)
                {
                    Velocity.x += sinYaw * PlayerConfig.Movement.SprintJumpForwardBoost;
                    Velocity.z += cosYaw * PlayerConfig.Movement.SprintJumpForwardBoost;
                }
                Grounded = false;
                _jumpArmed = false; // Disarm until grounded with jump released
            }

            if (InWater)
            {
                if (input.JumpHeld) Velocity.y += 0.08f;
                if (input.SneakHeld) Velocity.y -= 0.08f;
                Grounded = false;
            }

            // 7. Fall tracking
            if (!Grounded && _fallStartY == null)
            {
                _fallStartY = Position.y;
            }

            // 8. Collision resolution
            bool prevGrounded = Grounded;
            var result = _collision.ResolveMovement(ref Position, ref Velocity, CurrentHeight, PlayerConfig.Dimensions.Width, Sneaking && Grounded);
            Grounded = result.OnGround;

            // 9. Gravity & damping
            if (!Grounded && !InWater && !InLava)
            {
                Velocity.y -= PlayerConfig.Movement.Gravity;
                Velocity.y *= PlayerConfig.Movement.VerticalDrag;
            }
            else if (InWater)
            {
                Velocity.y -= 0.02f;
                Velocity.y *= 0.88f;
            }
            else
            {
                Velocity.y = 0f;
            }

            float friction = Grounded ? groundFriction : airFriction;
            Velocity.x *= friction;
            Velocity.z *= friction;

            // 10. Landing
            if (!prevGrounded && Grounded)
            {
                float fallDist = 0f;
                if (_fallStartY != null)
                {
                    fallDist = MathF.Max(0f, _fallStartY.Value - Position.y);
                }
                _fallStartY = null;
                FallDistance = fallDist;

                // When landing while holding jump, keep disarmed so continuous jumping does not occur
                if (input.JumpHeld)
                {
                    _jumpArmed = false;
                }

                OnLanded?.Invoke(fallDist, result.GroundBlock);
            }

            transform.position = Position;
        }

        public void Teleport(Vector3 newPos)
        {
            Position = newPos;
            PreviousPosition = newPos;
            transform.position = newPos;
            Velocity = Vector3.zero;
            _fallStartY = null;
            FallDistance = 0f;
            _jumpArmed = true;
            Grounded = true;
        }
    }
}
